core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java

```diff
@@ -0,0 +1,345 @@
+    }
+
+    private EventLoopState state(Endpoint endpoint, SessionProtocol sessionProtocol) {
+        if (endpoint.isGroup()) {
+            return groupState(endpoint, sessionProtocol);
+        }
+
+        final String firstTryHost;
+        final String secondTryHost;
+        if (endpoint.hasIpAddr()) {
+            final String ipAddr = endpoint.ipAddr();
+            assert ipAddr != null;
+            firstTryHost = ipAddr;
+            if (endpoint.isIpAddrOnly()) {
+                secondTryHost = null;
+            } else {
+                secondTryHost = endpoint.host();
+            }
+        } else {
+            firstTryHost = endpoint.host();
+            secondTryHost = null;
+        }
+        final int port = endpoint.hasPort() ? endpoint.port() : sessionProtocol.defaultPort();
+
+        if (!predefinedEndpointStates.isEmpty()) {
+            EventLoopState state = predefinedEndpointState(firstTryHost, port);
+            if (state != null) {
+                return state;
+            }
+
+            if (secondTryHost != null) {
+                state = predefinedEndpointState(secondTryHost, port);
+                if (state != null) {
+                    return state;
+                }
+            }
+        }
+
+        final boolean isHttp1 = isHttp1(sessionProtocol);
+        final StateKey firstKey = new StateKey(firstTryHost, port, isHttp1);
+        EventLoopState state = states.get(firstKey);
+        if (state != null) {
+            return state;
+        }
+
+        if (secondTryHost != null) {
+            final StateKey secondKey = new StateKey(secondTryHost, port, isHttp1);
+            state = states.get(secondKey);
+            if (state != null) {
+                return state;
+            }
+        }
+
+        return state(firstKey, sessionProtocol);
+    }
+
+    private EventLoopState state(StateKey stateKey, SessionProtocol sessionProtocol) {
+        return states.computeIfAbsent(
+                stateKey, unused -> EventLoopState.of(eventLoops, maxNumEventLoops(sessionProtocol), this));
+    }
+
+    private EventLoopState groupState(Endpoint endpoint, SessionProtocol sessionProtocol) {
+        final String groupName = endpoint.groupName();
+        final PredefinedEndpointState predefinedGroup = predefinedGroupStates.get(groupName);
+        if (predefinedGroup != null) {
+            return predefinedGroup.getOrCreate();
+        }
+
+        return state(new StateKey(groupName, sessionProtocol.defaultPort(), isHttp1(sessionProtocol)),
+                     sessionProtocol);
+    }
+
+    private int maxNumEventLoops(SessionProtocol sessionProtocol) {
+        return isHttp1(sessionProtocol) ? maxNumEventLoopsPerHttp1Endpoint
+                                        : maxNumEventLoopsPerEndpoint;
+    }
+
+    // TODO(minwoox) Use SessionProtocolNegotiationCache and enpoint to bring the protocol version when the
+    //               SessionProtocol is HTTP and HTTPS.
```

Comment:
Let me make another PR for this. 😉 
We use `SocketAddress` as a key in `SessionProtocolNegotiationCache ` but I need to use different key because we don't have the `SocketAddress` here. I think that will make this a little bit more complicated, so let me do that in another PR.